File: bot/skills/tree.py
import logging
from datetime import datetime
from typing import Any

from telegram import Update
from telegram.ext import Application, ContextTypes

from handlers import ChatCommandHandler

logger = logging.getLogger(__name__)

now = datetime.now()
year = now.year
if now.month < 12:
    year -= 1

AOC_LEADERBOARD_LINK = (
    f"https://adventofcode.com/{year}/leaderboard/private/view/458538"
)


App = Application[Any, Any, Any, Any, Any, Any]


def add_tree(app: App, handlers_group: int):
    logger.info("registering tree handlers")
    app.add_handler(ChatCommandHandler("tree", tree), group=handlers_group)


async def tree(update: Update, context: ContextTypes.DEFAULT_TYPE):
    text = (
        f"🎄🎄🎄 Присоединяйся к ежегодному решению елки! 🎄🎄🎄 \n"
        f"👉👉👉 https://adventofcode.com/ 👈👈👈 \n"
        f"😼😼😼 VLDC leaderboard: {AOC_LEADERBOARD_LINK} \n"
        f"Join Code: `458538-e2a0698b`"
    )

    if update.effective_chat is None:
        return
    await context.bot.send_message(update.effective_chat.id, text)

File: internal/skill/tree.go
package skill

import (
	"context"
	"fmt"
	"time"

	"github.com/go-telegram/bot"
	"github.com/go-telegram/bot/models"

	appbot "github.com/vldc-hq/vldc-bot/internal/bot"
)

const aocLeaderboardID = "458538"

func TreeSkill() appbot.Skill {
	return appbot.Skill{
		Name: "tree",
		Hint: "Advent of Code",
		Register: func(b *bot.Bot, _ *appbot.Deps) {
			b.RegisterHandler(bot.HandlerTypeMessageText, "/tree", bot.MatchTypePrefix, handleTree)
		},
	}
}

func handleTree(ctx context.Context, b *bot.Bot, update *models.Update) {
	if update.Message == nil {
		return
	}

	now := time.Now()
	year := now.Year()
	if now.Month() < time.December {
		year--
	}

	text := fmt.Sprintf(
		"🎄🎄🎄 Присоединяйся к ежегодному решению елки! 🎄🎄🎄 \n"+
			"👉👉👉 https://adventofcode.com/ 👈👈👈 \n"+
			"😼😼😼 VLDC leaderboard: https://adventofcode.com/%d/leaderboard/private/view/%s \n"+
			"Join Code: `458538-e2a0698b`",
		year, aocLeaderboardID,
	)

	_, _ = b.SendMessage(ctx, &bot.SendMessageParams{
		ChatID: update.Message.Chat.ID,
		Text:   text,
	})
}
